File: src/behavioural/memento/java/CareTaker.java
public class CareTaker {
    private final Stack<EditorMemento> history = new Stack<>();

    public void saveState(TextEditor editor) {
        // save the state
        history.push(editor.save())
    }

    public void undo(TextEditor editor) {
        if(!history.empty()) {
            history.pop()
            editor.restore(history.peek())
        }
    }
}
// 

File: src/behavioural/memento/ts/CareTaker.ts
class CareTaker {
  private history: EditorMemento[] = [];

  saveState(editor: TextEditor): void {
    this.history.push(editor.save());
  }

  undo(editor: TextEditor): void {
    if (this.history.length > 1) {
      this.history.pop();
      editor.restore(this.history[this.history.length - 1]);
    }
  }
}
